### PR TITLE
iconMargins mixing up fix

### PR DESCRIPTION
**Pull request by Katsz**
As of `2.0.1` `iconMargins` is not working (properly) because [here](https://github.com/arn00s/cariocamenu/blob/2.0.1/Sources/CariocaIndicatorView.swift#L180) constrains are assumed to be arranged in "CSS order": 
```swift
private func applyMarginConstraints(margins: (top: CGFloat, right: CGFloat, bottom: CGFloat, left: CGFloat)) {
	iconConstraints[0].constant = margins.top
	iconConstraints[1].constant = margins.right
	iconConstraints[2].constant = margins.bottom
	iconConstraints[3].constant = margins.left
	setNeedsLayout()
}
```
by in reality they are [arranged differently](https://github.com/arn00s/cariocamenu/blob/2.0.1/Sources/CariocaMenu%2BUtils.swift#L40):
```swift
func makeAnchorConstraints(to superview: UIView) -> [NSLayoutConstraint] {
	return [self.topAnchor.constraint(equalTo: superview.topAnchor),
		self.rightAnchor.constraint(equalTo: superview.rightAnchor),
		self.leftAnchor.constraint(equalTo: superview.leftAnchor),
		self.bottomAnchor.constraint(equalTo: superview.bottomAnchor)]
}
```
(note how `left` and `bottom` are misplaced)

Furthermore, (because of how iOS resolving constants in constraints depending on anchors order) `left` and `bottom`  should be negative to achieve desired result. 

So, this PR contains fix for those issues and cleanup of some duplicating code (addSubview, create constraints, addConstraints)

### CODE REVIEW

**Review by zigzagg16 — COMMENTED** [review]
Hey @Katsz ! First of all, thanks a lot for this PR!
It makes a lot of sense, I thought about using the EdgeInsets some time ago, but never did it ;)

Could you please use develop as a target ?
I'll add some comments below.

File: Sources/CariocaIndicatorConfiguration.swift [hunk 1]
```diff
@@ -27,12 +27,12 @@ public protocol CariocaIndicatorConfiguration {
 	///The custom shape of the view
 	func shape(for edge: UIRectEdge, frame: CGRect) -> UIBezierPath
 	///The margins for the icon, depending on the edge
-	func iconMargins(for edge: UIRectEdge) -> (top: CGFloat, right: CGFloat, bottom: CGFloat, left: CGFloat)
+	func iconMargins(for edge: UIRectEdge) -> UIEdgeInsets
 }
 public extension CariocaIndicatorConfiguration {
 	///Default margins are 0,0,0,0
```
Comment (zigzagg16) [inline Sources/CariocaIndicatorConfiguration.swift:33]: The default values were 0,0,0,0. Is it still the same ?

File: Sources/CariocaIndicatorView.swift [hunk 2]
```diff
@@ -39,7 +39,7 @@ public class CariocaIndicatorView: UIView {
 	///The custom indicator configuration
 	private let config: CariocaIndicator
 	///The constraints applied to the iconview. Can be updated later with custom configuration
-	private var iconConstraints: [NSLayoutConstraint] = []
+	private var iconConstraints: EdgeConstraints!
```
Comment (zigzagg16) [inline Sources/CariocaIndicatorView.swift:42]: Is there a way we can avoid this force-unwrap ?

File: Sources/EdgeConstraints.swift [hunk 3]
```diff
@@ -0,0 +1,27 @@
+//
+// Created by Alex Katsz on 2019-03-11.
+// Copyright (c) 2019 CariocaMenu. All rights reserved.
+//
+
+import UIKit
+
+public struct EdgeConstraints {
```
Comment (zigzagg16) [inline Sources/EdgeConstraints.swift:9]: Could you please add comments here for methods and properties ? I know some comment may seem useless in some cases, but I like to keep a 100% documentation coverage, as it's public code.
You can run jazzy to update the generated docs.

File: Sources/CariocaIndicatorConfiguration.swift (hunk 1 above)
Comment (Katsz) [reply Sources/CariocaIndicatorConfiguration.swift:33]: [Yeah](https://developer.apple.com/documentation/uikit/uiedgeinsets/1625050-init)

File: Sources/CariocaIndicatorView.swift (hunk 2 above)
Comment (Katsz) [reply Sources/CariocaIndicatorView.swift:42]: Well, take a look at `init(edge: UIRectEdge, indicator: CariocaIndicator)`, only if we can add (or, at least create) constraints before calling super.init(frame: frame).

Another way I can think of - we can make it `lazy var` and call after `init` is completed, but it sounds messy.

File: Sources/EdgeConstraints.swift (hunk 3 above)
Comment (Katsz) [reply Sources/EdgeConstraints.swift:9]: Which jazzy version you are using? `0.9.5` complaining about `CariocaMenu+Protocols.swift#12`
```
"symbol_kind": "source.lang.swift.decl.typealias",
"warning": "undocumented"
```

**Comment by Katsz** [comment]
I ended up adding some comments, could you double-check them? I'm not native speaker as you may have already notice `¯\_(ツ)_/¯`. 
I cant think about descriptive comment for `typealias CariocaController`, so atm only 99% cov.

**Comment by Katsz** [comment]
Could you by the way explain why `typealias CariocaIndicator = UIView & CariocaIndicatorConfiguration` rather then just `CariocaIndicatorConfiguration`?  There is no single usage of `UIView` part.

And is there any reason to `didSelectItem(at: Int)` in `CariocaMenu` uses `0.5`,  as animation timing, then applying some math lilke
```swift
var timingAnim1: Double = firstStepDuration * 0.7, timingAnim2: Double = firstStepDuration * 0.3
if hasBoomerang {
    // ...
    timingAnim1 = firstStepDuration * 1.25
}
```  
instead of fetching it outside (i.e. from `CariocaController`)?  That will be more flexible solution

File: Sources/CariocaIndicatorView.swift (hunk 2 above)
Comment (zigzagg16) [reply Sources/CariocaIndicatorView.swift:42]: right, then keep it as it is :)

File: Sources/EdgeConstraints.swift (hunk 3 above)
Comment (zigzagg16) [reply Sources/EdgeConstraints.swift:9]: THanks! Just updated the docs. I'm using 0.9.5 too :)

**Comment by zigzagg16** [comment]
Hey @Katsz , looks like the docs generated conflicts. Can you fix those ? As soon as it's done, I'll merge the PR.
Thanks a lot for your work, one more time ! 💯

**Comment by Katsz** [comment]
It seems like your `develop` is ahead of `master`, including swift version migration, copyright removal, etc. I'm considering closing PR and re-opening it from `katsz:develop` to `arn00s:develop` to make git history less spaghetti-ish.

**Pull request closed without merging**